### PR TITLE
connectors-ci: improve pytest result evaluation

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Relates to #27425 , #28423

### Original problem
When running pytest in a dagger container we face two challenges:
* When a pytest suite is empty or all tests are skipped pytest will return a 5 exit code. exit code` >0` in Dagger raise ExecError, but we want to mark this step as skipped and not fail.
* When CAT (which is a pytest suite) is failing still need to perform after the facts operations to retrieve updated connector configurations from the container. When a container raises an ExecError we can't interact with it's filesystem without raising the same ExecError again.

#### How we overcame this problem so far
We installed a pytest plugin to always make pytest suite exit with code `0`: `pytest-custom_exit_code`. We inferred success / failure / skipping of test by parsing the stdout log. This is not robust as the stdout content can change according to the pytest version or if a CAT run outputs unexpected stuff to stdout. **This led to the false positive errors mentioned in #27425**

#### The new approach to overcome this problem
We still make CAT exec artificially exit with a `0` exit code, but we store the original exit code to a file in the container filesystem: `/exit_code`. We `get_step_result` is run it will check if such a file exist on the container and return the in file exit code value instead of the pure container exit code. This allows us to capture the actual exit code that pytest returned.   

## How
The new approach mentioned above is mainly introduced via:
- a new hack: `hacks.never_fail_exec` function:
https://github.com/airbytehq/airbyte/blob/ddb7345e008c7e3c24db77a04bc6a9c45855f528/airbyte-ci/connectors/pipelines/pipelines/hacks.py#L145
- updating the `Step.get_step_result` function to read in file exit code if `/exit_code` file is available on the container.
https://github.com/airbytehq/airbyte/blob/ddb7345e008c7e3c24db77a04bc6a9c45855f528/airbyte-ci/connectors/pipelines/pipelines/bases.py#L245

**I tried to thoroughly unit test these change to show how async test with a dagger client work:**
https://github.com/airbytehq/airbyte/blob/ddb7345e008c7e3c24db77a04bc6a9c45855f528/airbyte-ci/connectors/pipelines/tests/tests/test_common.py#L20

### Centralizing the AcceptanceTest logic
The logic to build the CAT container was in the  `environment.with_connector_acceptance_test` function. I decided to move it to `common.AcceptanceTest._build_connector_acceptance_test` as `environment` functions are originally made to be reused in multiple steps. This function is very specific to `AcceptanceTest`.

### Make all step write the `stdout` / `stderr` to a log file
(Relates to #28423)
We originally only wrote CAT output logs to a local directory. I figured that this logic can easily apply to all steps and would provide more debugging abilities to developers. Logs are written when the step results are built in `get_step_results`
https://github.com/airbytehq/airbyte/blob/ddb7345e008c7e3c24db77a04bc6a9c45855f528/airbyte-ci/connectors/pipelines/pipelines/bases.py#L258

### Wrap acceptance test into its own dagger `Pipeline`.
(Relates to #28423)
I created Step specific dagger client in `AcceptanceTest` in order to showcase how we could make step execution wrapped into a dagger `Pipeline`: this will make the CAT dagger operations grouped in a explicit way in the dagger UI.
https://github.com/airbytehq/airbyte/blob/ddb7345e008c7e3c24db77a04bc6a9c45855f528/airbyte-ci/connectors/pipelines/pipelines/bases.py#L112

## Recommended reading order
1. https://github.com/airbytehq/airbyte/blob/ddb7345e008c7e3c24db77a04bc6a9c45855f528/airbyte-ci/connectors/pipelines/pipelines/tests/common.py#L176
2. https://github.com/airbytehq/airbyte/blob/ddb7345e008c7e3c24db77a04bc6a9c45855f528/airbyte-ci/connectors/pipelines/tests/tests/test_common.py#L20

### CODE REVIEW

**Comment by octavia-squidington-iii** [comment]
##  source-pokeapi test report (commit [`9ae813c9e4`](https://github.com/airbytehq/airbyte/pull/28767/commits/9ae813c9e4918b407c2ba168bde3c3d588ccfb41)) - ❌

⏲️  Total pipeline duration: 02mn32s 

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-pokeapi/metadata.yaml | ✅        |
| Connector version semver check                                        | ✅        |
| QA checks                                                             | ✅        |
| Code format checks                                                    | ✅        |
| Connector package install                                             | ✅        |
| Build source-pokeapi docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                            | ✅        |
| Acceptance tests                                                      | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/augustin_connectors-ci_change-pytest-step-result-evaluation/1690444083/9ae813c9e4918b407c2ba168bde3c3d588ccfb41/source-pokeapi/0.1.5/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-pokeapi test
```

**Comment by octavia-squidington-iii** [comment]
##  source-openweather test report (commit [`9ae813c9e4`](https://github.com/airbytehq/airbyte/pull/28767/commits/9ae813c9e4918b407c2ba168bde3c3d588ccfb41)) - ✅

⏲️  Total pipeline duration: 03mn46s 

| Step                                                                      | Result   |
|:--------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-openweather/metadata.yaml | ✅        |
| Connector version semver check                                            | ✅        |
| QA checks                                                                 | ✅        |
| Code format checks                                                        | ✅        |
| Connector package install                                                 | ✅        |
| Build source-openweather docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                | ✅        |
| Acceptance tests                                                          | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/augustin_connectors-ci_change-pytest-step-result-evaluation/1690444083/9ae813c9e4918b407c2ba168bde3c3d588ccfb41/source-openweather/0.1.6/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-openweather test
```

File: airbyte-ci/connectors/pipelines/pipelines/tests/common.py [hunk 1]
```diff
@@ -189,19 +204,57 @@ async def _run(self, connector_under_test_image_tar: Optional[File]) -> StepResu
...
+        This bursts the CAT cached results everyday.
+        It's cool because in case of a partially failing nightly build the connectors that already ran CAT won't re-run CAT.
+        We keep the guarantee that a CAT runs everyday.
+
+        Returns:
+            str: A string representing the current date.
+        """
+        return datetime.datetime.utcnow().strftime("%Y%m%d")
+
+    async def _build_connector_acceptance_test(self, connector_under_test_image_tar: File) -> Container:
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/tests/common.py:231]: well done function.

But Im starting to notice we have a lot of connector code, that isnt obvious based on the file structure.

For example this is in test/common

should we start making it more specific. e.g.:
- test/connector_acceptance/common.py
- actions/connectors/tests.py

**Review by bnchrch — COMMENTED** [review]
A few comments but the only one I think is a hard blocker is referencing context.connector in `Step`

File: .github/workflows/airbyte_ci_tests.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+name: Airbyte CI pipeline tests
+
+on:
+  push:
+    branches:
+      - !master
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply .github/workflows/airbyte_ci_tests.yml:6]: Sorry, it was a WIP push from a different branch: https://github.com/airbytehq/airbyte/pull/28857

File: airbyte-ci/connectors/pipelines/pipelines/tests/common.py (hunk 1 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/tests/common.py:231]: The `test/common` module is originally made to receive connector testing step that are agnostic to the connector language. This is why `AcceptanceTest` belongs here.
The `actions` module was originally made to receive re-usable code. But I think I overused it and suffed action/environements with code that is not re-used in multiple context.

**Review by bnchrch — APPROVED** [review]
lgtm!

**Pull request merged**